engine/src/test/java/io/camunda/zeebe/engine/processing/timer/TimerStartEventTest.java

```diff
@@ -1105,21 +1105,26 @@ public void shouldTriggerOnlyTwice() {
             firstDeploymentProcessDefinitionKey,
             secondDeploymentProcessDefinitionKey);
 
-    // due timers are only checked if the engine is not currently processing #10112
-    Awaitility.await("until the engine is idle again").until(engine::hasReachedEnd);
-
     // when
     engine.increaseTime(Duration.ofSeconds(10));
 
     // then
-    assertThat(RecordingExporter.timerRecords(TimerIntent.TRIGGERED).limit(5))
-        .extracting(record -> record.getValue().getProcessDefinitionKey())
-        .containsExactly(
-            firstDeploymentProcessDefinitionKey,
-            secondDeploymentProcessDefinitionKey,
-            firstDeploymentProcessDefinitionKey,
-            secondDeploymentProcessDefinitionKey,
-            secondDeploymentProcessDefinitionKey);
+    Awaitility.await()
+        .untilAsserted(
+            () -> {
+              // due timers are only checked if the engine is not currently processing #10112
+              // so we may need to move the clock slightly for each check
+              engine.increaseTime(Duration.ofMillis(100));
+              assertThat(RecordingExporter.timerRecords(TimerIntent.TRIGGERED).limit(5))
+                  .describedAs("Expect that start_1 triggered twice and start_2 triggered thrice")
+                  .extracting(record -> record.getValue().getProcessDefinitionKey())
+                  .containsExactly(
+                      firstDeploymentProcessDefinitionKey,
+                      secondDeploymentProcessDefinitionKey,
+                      firstDeploymentProcessDefinitionKey,
+                      secondDeploymentProcessDefinitionKey,
+                      secondDeploymentProcessDefinitionKey);
+            });
```

Comment:
❓ In the commit you mention
> So, if it only triggered 4 times at the
moment we check it, then we have to wait 5 seconds until we can increase
the time.

Isn't this the case for the previous solution as well? The only downside in the old solution is that it would fail the test after waiting the 5 seconds, here it will try it a second time before Awaitility gives a timeout (10s).

Reply:
>The only downside in the old solution is that it would fail the test after waiting the 5 seconds, here it will try it a second time before Awaitility gives a timeout (10s).

No, the downside of the old solution was that it depends on a technical concept (that the engine must be idle in other for the DueDateTimerChecker to check for expired timers). The solution provided in 6bee15abe3760811d6722890f1b6aeb2b069a004 didn't have that problem but came with the downside of being slow (taking 5 seconds or so). That's why a later commit was used to improve the performance of the test.